chaos/lock-rpc-server.go

```diff
@@ -100,17 +115,130 @@ func (l *lockServer) RUnlock(args *dsync.LockArgs, reply *bool) error {
 	}
-	if locksHeld > ReadLock {
-		l.lockMap[args.Name] = locksHeld - ReadLock // Remove one of the read locks held
-	} else {
-		delete(l.lockMap, args.Name) // Remove the (last) read lock
+	if l.removeEntry(args.Name, args.Uid, &lri) {
+		return nil
+	}
+	return fmt.Errorf("RUnlock unable to find corresponding read lock for uuid: %s", args.Uid)
+}
+
+func (l* lockServer) Active(args *dsync.LockArgs, reply *bool) error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if err := l.verifyArgs(args); err != nil {
+		return err
+	}
+	var lri []lockRequesterInfo
+	if lri, *reply = l.lockMap[args.Name]; !*reply {
+		return nil // No lock is held on the given name so return false
+	}
+	// Check whether uid is still active
+	for _, entry := range lri {
+		if *reply = entry.uid == args.Uid; *reply {
+			return nil // When uid found return true
+		}
+	}
+	return nil // None found so return false
+}
+
+// removeEntry either, based on the uid of the lock message, removes a single entry from the
+// lockRequesterInfo array or the whole array from the map (in case of a write lock or last read lock)
+func (l *lockServer) removeEntry(name, uid string, lri *[]lockRequesterInfo) bool {
+	// Find correct entry to remove based on uid
+	for index, entry := range *lri {
+		if entry.uid == uid {
+			if len(*lri) == 1 {
+				delete(l.lockMap, name) // Remove the (last) lock
+			} else {
+				// Remove the appropriate read lock
+				*lri = append((*lri)[:index], (*lri)[index+1:]...)
+				l.lockMap[name] = *lri
+			}
+			return true
+		}
+	}
+	return false
+}
+
+type nameLockRequesterInfoPair struct {
+	name string
+	lri  lockRequesterInfo
+}
+
+// getLongLivedLocks returns locks that are older than a certain time and
+// have not been 'checked' for validity too soon enough
+func getLongLivedLocks(m map[string][]lockRequesterInfo, interval time.Duration) []nameLockRequesterInfoPair {
+
+	rslt := []nameLockRequesterInfoPair{}
+
+	for name, lriArray := range m {
+
+		for idx, _ := range lriArray {
+			// Check whether enough time has gone by since last check
+			if time.Since(lriArray[idx].timeLastCheck) >= interval {
+				rslt = append(rslt, nameLockRequesterInfoPair{name: name, lri: lriArray[idx]})
+				lriArray[idx].timeLastCheck = time.Now()
+			}
+		}
+	}
+
+	return rslt
+}
+
+// lockMaintenance loops over locks that have been active for some time and checks back
+// with the original server whether it is still alive or not
+func (l *lockServer) lockMaintenance(interval time.Duration) {
+
+	l.mutex.Lock()
+	// get list of locks to check
+	nlripLongLived := getLongLivedLocks(l.lockMap, interval)
+	l.mutex.Unlock()
+
+	for _, nlrip := range nlripLongLived {
+
+		c := newClient(nlrip.lri.node, nlrip.lri.rpcPath)
+
+		var active bool
+
+		// Call back to original server verify whether the lock is still active (based on name & uid)
+		if err := c.Call("Dsync.Active", &dsync.LockArgs{Name: nlrip.name, Uid: nlrip.lri.uid}, &active); err != nil {
```

Comment:
Good thinking, however you wouldn't want to use RLock since they will always be granted when another RLock for the same name is already in place, so that doesn't tell you whether the lock is available again or not.

Initial version used Lock which works (eventually) but imagine a busy resource/name with many read locks on it in the distributed system (either simultaneous or sequentially). Then there is a big chance that (during maintenance that kicks in every minute or so) the Lock would fail due to a new RLock already being in place, and hence the stale lock would not be detected.
You can even think of a scenario where multiple RLock are 'overlapping' in time meaning that the count is always >= 1, and the Lock cannot be granted at all (or would take a really long time). 

As the `uid` was added anyway, the `Active` check is a better way for detecting that a lock is stale that also caters for the cases above. Another advantage is that there is no need to send an Unlock message which can fail, which then leads to another stale lock at the other end (small chance, but still).